optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/JaxbIO.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.io.jaxb;
+
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMResult;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import org.optaplanner.core.impl.io.XmlIO;
+import org.optaplanner.core.impl.io.XmlUnmarshallingException;
+
+public final class JaxbIO<T> implements XmlIO<T> {
+    private static final int DEFAULT_INDENTATION = 2;
+
+    private final Unmarshaller unmarshaller;
+    private final Marshaller marshaller;
+    private final Class<T> rootClass;
+    private final int indentation;
+
+    public JaxbIO(Class<T> rootClass) {
+        this(rootClass, DEFAULT_INDENTATION);
+    }
+
+    public JaxbIO(Class<T> rootClass, int indentation) {
+        if (rootClass == null) {
+            throw new IllegalArgumentException("Root element class cannot be null.");
+        }
+        this.rootClass = rootClass;
+        this.indentation = indentation;
+        try {
+            JAXBContext jaxbContext = JAXBContext.newInstance(rootClass);
+            unmarshaller = jaxbContext.createUnmarshaller();
+            marshaller = jaxbContext.createMarshaller();
+            marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
+            marshaller.setProperty(Marshaller.JAXB_ENCODING, StandardCharsets.UTF_8.toString());
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format("Unable to create JAXB marshaller or unmarshaller for a root element class %s.",
+                    rootClass.getName());
+            throw new IllegalStateException(errMessage, jaxbException);
+        }
+    }
+
+    public T read(Reader reader) {
+        if (reader == null) {
+            throw new IllegalArgumentException("Reader cannot be null.");
+        }
```

Comment:
Personally, I think that `Objects.requireNonNull()` at use site would sufficehere. You're just replacing one exception with another.